bin/get-amy.py: proper argparsing, args for tag filtering

pbanaszkiewicz opened the pull request:
1. Add proper solution for argument parsing (argparse module from
   Python).
2. Add '-t'/'--tag' arguments for specifying tags used in event
   filtering.
3. Filter events (via query) by tags provided.

pbanaszkiewicz committed:
bin/get-amy.py: proper argparsing, args for tag filtering

1. Add proper solution for argument parsing (argparse module from
Python).
2. Add '-t'/'--tag' arguments for specifying tags used in event
filtering.
3. Filter events (via query) by tags provided.

gvwilson commented on bin/get-amy.py:
```diff
@@ -82,4 +89,17 @@ def sort_flags(data):
 
 if __name__ == '__main__':
-    main()
+    parser = argparse.ArgumentParser(description=sys.modules[__name__].__doc__)
```
I would put all the argument handling code in a function that returns a triple (url, output_file, tags).

gvwilson replied:
Does commands.mk have to be updated to match the new parameter syntax?

pbanaszkiewicz replied:
And where that file is located?

gvwilson replied:
which file?

pbanaszkiewicz replied:
`commands.mk` you mentioned before

gvwilson replied:
Um... it's in the unified-styles branch in my copy of the repo. Please 
add it to 'Makefile' for now and I'll figure it out in the merge - thanks.

gvwilson commented on bin/get-amy.py:
```diff
@@ -38,15 +35,25 @@ def main():
+    try:
+        with open(output_file, 'w') as writer:
+            yaml.dump(config, writer, encoding='utf-8', allow_unicode=True)
+    except TypeError:
```
Uh, really? Wouldn't it be better to test output_file to see if it's string-like using a plain old 'if'?

pbanaszkiewicz replied:
While Python folks suggest using duck-typing, I must admit I don't know what errors can `yaml.dump` throw (maybe even some TypeErrors?), so I will change that to the if-case.

gvwilson commented on bin/get-amy.py:
```diff
@@ -38,15 +35,25 @@ def main():
+    # Consider filtering by tags.
+    if tags:
+        query_params = [urllib.parse.urlencode({'tag': tag}) for tag in tags]
+        query = '&'.join(query_params)
```
Isn't there a utility function in urllib that takes a dict and construct the full URL?  Or am I mis-remembering? (Seems odd to have to do the '&' and '?' ourselves.

wking replied:
On Fri, Jan 01, 2016 at 02:01:56PM -0800, Greg Wilson wrote:

> > -        query_params = [urllib.parse.urlencode({'tag': tag}) for tag in tags]
> > -        query = '&'.join(query_params)
> 
> Isn't there a utility function in urllib that takes a dict and
> construct the full URL?  Or am I mis-remembering? (Seems odd to have
> to do the '&' and '?' ourselves.

I'd rather not have multiple tag=… entries in one string.  Can't we
do:

  tag=a,b,c

with:

  query = urllib.parse.urlencode({'tag': ','.join(tags)})

pbanaszkiewicz replied:
Unfortunately we can't do this.
1. AMY uses `?tag=SWC&tag=DC&host=123` form, and…
2. multiple items of the same name are Django's way of handling forms (see
   for example AMY's code for bulk upload) and GET query params: instead of
   `dict`, Django creates a `QueryDict`, which provides `.getlist()` method

@gvwilson: sadly urlencode doesn't work with `{'tag': ['SWC', 'DC']}` - it
makes literal URL encoded string representation of the list `['SWC', 'DC']`.

On Fri, Jan 1, 2016 at 11:09 PM, W. Trevor King notifications@github.com
wrote:

> In bin/get-amy.py
> https://github.com/swcarpentry/website/pull/118#discussion_r48680717:
> 
> > ```
> >  '''Download and save data.'''
> > 
> >  address = base_url + url
> > ```
> > -    # Consider filtering by tags.
> > -    if tags:
> > -        query_params = [urllib.parse.urlencode({'tag': tag}) for tag in tags]
> > -        query = '&'.join(query_params)
> 
> On Fri, Jan 01, 2016 at 02:01:56PM -0800, Greg Wilson wrote: > +
> query_params = [urllib.parse.urlencode({'tag': tag}) for tag in tags] > +
> query = '&'.join(query_params) Isn't there a utility function in urllib
> that takes a dict and construct the full URL? Or am I mis-remembering?
> (Seems odd to have to do the '&' and '?' ourselves.
> I'd rather not have multiple tag=… entries in one string. Can't we do:
> tag=a,b,c with: query = urllib.parse.urlencode({'tag': ','.join(tags)})
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/swcarpentry/website/pull/118/files#r48680717.

gvwilson replied:
What's wrong with multiple tag=XYZ in one query string?  If the 
receiving end does the right thing, that seems better than requiring 
another level of parsing.

wking replied:
On Fri, Jan 01, 2016 at 02:20:12PM -0800, Piotr Banaszkiewicz wrote:

> 1. multiple items of the same name are Django's way of handling
>    forms (see for example AMY's code for bulk upload) and GET query
>    params: instead of `dict`, Django creates a `QueryDict`, which
>    provides `.getlist()` method

You don't have to use .getlist, you can use something like:

  tags = request.GET.get('tag', '').split(',')

But without an AMY change like that, I don't see a way around the ‘&’
joining you're using here.  In the end, it doesn't really matter which
way you go.

gvwilson replied:
Thanks - who knew the Internet was so complicated? :-)

pbanaszkiewicz replied:
There's a solution on the horizon:

``` python
>>> query = [('tag', 'SWC'), ('tag', 'DC')]
>>> urlencode(query)
'tag=SWC&tag=DC'
```

pbanaszkiewicz committed:
bin/get-amy.py: move argparsing to separate function

pbanaszkiewicz committed:
bin/get-amy.py: switch duck-typing to if-case

...when checking if provided var is a file or a name of file to be
opened.

pbanaszkiewicz committed:
bin/get-amy.py: replace ugly query param generation with nice one

We needed to create an URL-encoded query string containing multiple
identical keys:
`key=A&key=B`

Fortunately urlencode supports not only dicts, but also sequence of
pairs, which I happily switched to.

pbanaszkiewicz committed:
Makefile: fix args in invocation of bin/get-amy.py
